File: jsmol_bokeh_extension/jsmol.py
"""JSMol bokeh extension - Python extension"""
from os import path
from bokeh.core.properties import Instance, String, Dict
from bokeh.models import ColumnDataSource, LayoutDOM
from bokeh.util.compiler import TypeScript

directory = path.dirname(path.realpath(__file__))
with open(path.join(directory, 'jsmol.ts'), 'r', encoding='utf8') as f:
    TS_CODE = f.read()


# This custom extension model will have a DOM view that should layout-able in
# Bokeh layouts, so use ``LayoutDOM`` as the base class. If you wanted to create
# a custom tool, you could inherit from ``Tool``, or from ``Glyph`` if you
# wanted to create a custom glyph, etc.
class JSMol(LayoutDOM):  # pylint: disable=too-few-public-methods
    """Extension for visualizing atomic structues via JSMol."""

    # The special class attribute ``__implementation__`` should contain a string
    # of JavaScript (or CoffeeScript) code that implements the JavaScript side
    # of the custom extension model.
    __implementation__ = TypeScript(TS_CODE)

    # Below are all the "properties" for this model. Bokeh properties are
    # class attributes that define the fields (and their types) that can be
    # communicated automatically between Python and the browser. Properties
    # also support type validation. More information about properties in
    # can be found here:
    #
    #    https://bokeh.pydata.org/en/latest/docs/reference/core.html#bokeh-core-properties

    # This is a Bokeh Dict providing the JSMol "Info variable"
    # See http://wiki.jmol.org/index.php/Jmol_JavaScript_Object/Info
    info = Dict(String, String)

    # Currently (mis)using this to pass a script to be executed since there
    # doesn't seem to be a better solution (?)
    # https://groups.google.com/a/continuum.io/forum/#!topic/bokeh/0m17mNMnTys
    script_source = Instance(ColumnDataSource)

    # Path to JSMol javascript file (can be full or relative URL), e.g.
    # e.g. https://chemapps.stolaf.edu/jmol/jsmol/JSmol.min.js
    # or jsmol/JSmol.min.js
    js_url = String()

File: jsmol_bokeh_extension/jsmol.ts
// Model wrapping JSmol.
//
// "jsmol.py" contains the python counterpart.
declare var Jmol: any;
declare var jmolApplet0: any;
 

// These "require" lines are similar to python "import" statements
import {LayoutDOM, LayoutDOMView} from "models/layouts/layout_dom"
import {LayoutItem} from "core/layout"
import {ColumnDataSource} from "models/sources/column_data_source"
import * as p from "core/properties"


// This defines some default options for JSmol
// See https://gist.github.com/jhjensen2/4701339 for more details.
const INFO = {
  height: "100%",
  width: "100%",
  serverURL: "https://chemapps.stolaf.edu/jmol/jsmol/php/jsmol.php",
  use: "HTML5",
  j2sPath: "https://chemapps.stolaf.edu/jmol/jsmol/j2s",
  script: "background black;load https://dev-www.materialscloud.org/cofs/api/v2/cifs/febd2d02-5690-4a07-9013-505c9a06bc5b/content/download",
  //disableJ2SLoadMonitor: true,
}

// To create custom model extensions that will render on to the HTML canvas
// or into the DOM, we must create a View subclass for the model.
//
// In this case we will subclass from the existing BokehJS ``LayoutDOMView``
export class JSMolView extends LayoutDOMView {
  model: JSMol
  private _applet: any

  initialize(): void {
      super.initialize()
  
      var url = this.model.js_url
      if (! url) {
        url = "https://chemapps.stolaf.edu/jmol/jsmol/JSmol.min.js"
      }
      const script = document.createElement('script')
      script.src = url
      script.async = false
      script.onload = () => this._init()
      //script.onreadystatechange = (script.onload = () => this._init())
      //script.onreadystatechange = script.onload
      document.querySelector("head")!.appendChild(script)
  }
     
  private _init(): void {
    // Create a new Jmol applet using the JSmol.js API. This assumes JSmol.js
    // has already been loaded.
    // Models will be able to specify and load external scripts automatically.
    //
    // BokehJS Views create <div> elements by default, accessible as this.el.

    // if no info dict provided, use reasonable default
    if (! this.model.info) {
      this.model.info = INFO
    }

    // disable usage tracker - this conflicts with cross-site-scripting policies when served over https
    delete Jmol._tracker

    // returns html + assigns applet object to "jmolApplet0" variable
    var html = Jmol.getAppletHtml("jmolApplet0", this.model.info)
    this._applet = jmolApplet0
    this.el.innerHTML = html

    // avoid creating a "deferred" applet
    // https://sourceforge.net/p/jsmol/discussion/general/thread/48083aa7///10a6/bc1c
    this._applet._cover(false)

    // Set a listener so that when the Bokeh script input changes it is executed
    this.connect(this.model.script_source.properties.data.change, () => {
        console.log("Script source changed")
        Jmol.script(this._applet, this.model.script_source.get_column('script')![0])
    })
  }

  get child_models(): LayoutDOM[] {
    return []
  }

  _update_layout(): void {
    this.layout = new LayoutItem()
    this.layout.set_sizing(this.box_sizing())
  }
}


// We must also create a corresponding JavaScript BokehJS model subclass to
// correspond to the python Bokeh model subclass. In this case, since we want
// an element that can position itself in the DOM according to a Bokeh layout,
// we subclass from ``LayoutDOM``
export namespace JSMol {
    export type Attrs = p.AttrsOf<Props>
        
    export type Props = LayoutDOM.Props & {
        js_url: p.Property<string>,
        script_source: p.Property<ColumnDataSource>,
        info: p.Property<any>
    }
}


export interface JSMol extends JSMol.Attrs {}

export class JSMol extends LayoutDOM {
    properties: JSMol.Props

    constructor(attrs?: Partial<JSMol.Attrs>) {
        super(attrs)
    }

    static initClass() {
     // The ``type`` class attribute should generally match exactly the name
    // of the corresponding Python class.
    this.prototype.type = "JSMol"

     // This is usually boilerplate. In some cases there may not be a view.
    this.prototype.default_view = JSMolView


    // The define block adds corresponding "properties" to the JS model. These
    // should basically line up 1-1 with the Python model class. Most property
    // types have counterparts, e.g. ``bokeh.core.properties.String`` will be
    // ``p.String`` in the JS implementation. Where the JS type system is not yet
    // as rich, you can use ``p.Any`` as a "wildcard" property type.
    this.define<JSMol.Props>({
      script_source: [ p.Instance         ],
      info:          [ p.Any              ],
      js_url:        [ p.String           ],
    })
  }
}
JSMol.initClass()
